robustifier/src/main/kotlin/cmu/isr/tolerance/Utils.kt

```diff
@@ -0,0 +1,139 @@
+import cmu.isr.ts.LTS
+import cmu.isr.ts.MutableDetLTS
+import cmu.isr.ts.MutableLTS
+import cmu.isr.ts.alphabet
+import cmu.isr.ts.lts.CompactDetLTS
+import cmu.isr.ts.lts.asLTS
+import cmu.isr.ts.lts.CompactLTS
+import cmu.isr.ts.lts.checkSafety
+import cmu.isr.ts.lts.makeErrorState
+import cmu.isr.ts.nfa.determinise
+import net.automatalib.automata.fsa.impl.compact.CompactDFA
+import net.automatalib.util.automata.builders.AutomatonBuilders
+import net.automatalib.words.Alphabet
+import java.util.*
+
+/**
+ * Finds all transitions in T
+ */
+fun ltsTransitions(T : LTS<Int, String>) : Set<Triple<Int,String,Int>> {
+    val init = T.initialStates
+    if (init.isEmpty()) {
+        return emptySet()
+    }
+
+    // BFS for all transitions
+    val transitions = mutableSetOf<Triple<Int,String,Int>>()
+    val visited : MutableSet<Int> = mutableSetOf()
+    val q : Queue<Int> = LinkedList()
+    q.addAll(init)
+    while (q.isNotEmpty()) {
+        val state = q.remove()
+        if (visited.contains(state)) {
+            continue
+        }
+        visited.add(state)
+        for (a in T.alphabet()) {
+            val adjacent = T.getTransitions(state, a)
+            for (dst in adjacent) {
+                transitions.add(Triple(state, a, dst))
+                q.add(dst)
+            }
+        }
+    }
+
+    return transitions
+}
+
+/**
+ * It's ridiculous how tough they make it to copy.
+ * The algorithm makes a key assumption: that initial states are the lowest state ID's, and will be added to newLTS
+ * with the same lowest state IDs.
+ */
+fun copyLTS(T : CompactDetLTS<String>) : CompactDetLTS<String> {
```

Comment:
Not sure why you need this. Doesn't the TSCopy.copy function work here?

Reply:
I didn't realize there was a copy function, thanks!

Reply:
Woah I may need a bit of a primer on how to use TSCopy.copy, the function looks quite intimidating: https://learnlib.github.io/automatalib/maven-site/0.7.0/apidocs/net/automatalib/util/ts/copy/TSCopy.html

Reply:
It turns out TSCopy.copy shuffles the state IDs around. This is unacceptable for my use case, unless there's a work around that I'm missing. I realize that copy() returns a mapping from original states to new states (this is how I was able to confirm that shuffling is going on), but I don't think this helps.

Unless there's a way to define a custom copy visitor that doesn't shuffle the state IDs around?

Reply:
I think you could. TSCopy uses TSTraverse visitors to traverse the model. But then, it is similar to what you are doing right now.

I quickly go through Automatalib. I think you could try the CompactDFA(other) constructor. It takes the another DFA and returns a new one. It seems to make a copy of the transitions. You could do an experiment.

Reply:
This was a great idea, but it turns out that the copy constructor is protected for CompactNFA:

http://learnlib.github.io/automatalib/maven-site/latest/apidocs/net/automatalib/automata/fsa/impl/compact/CompactNFA.html#CompactNFA-net.automatalib.words.Alphabet-net.automatalib.automata.fsa.impl.compact.CompactNFA-

Reply:
Wait, that's a little weird. Because I think I have used it somewhere in my implementation. You could try it out. Maybe they updated the code but does not update the document.